### PR TITLE
Implement CharmStore.Get.

### PR DESCRIPTION
**Pull request by frankban**
CharmStore.Get is used to retrieve a charm
from the charm store.

### CODE REVIEW

File: charmrepo/charmstore.go [hunk 1]
```diff
@@ -63,11 +69,65 @@ func NewCharmStore(p NewCharmStoreParams) (Interface, error) {
 	}, nil
 }
 
-var notImplemented = errgo.New("not implemented")
-
 // Get implements Interface.Get.
 func (s *CharmStore) Get(curl *charm.URL) (charm.Charm, error) {
-	return nil, notImplemented
+	// Disable bundle deployment.
```
Comment (rogpeppe) [inline charmrepo/charmstore.go]: This comment seems somewhat redundant to me

File: charmrepo/charmstore.go [hunk 2]
```diff
@@ -63,11 +69,65 @@ func NewCharmStore(p NewCharmStoreParams) (Interface, error) {
...
+	if curl.Series == "bundle" {
+		return nil, errgo.Newf("expected a charm URL, got bundle URL %q", curl)
+	}
+
+	// Prepare the cache directory and retrieve the charm.
+	if err := os.MkdirAll(s.cacheDir, os.FileMode(0755)); err != nil {
```
Comment (rogpeppe) [inline charmrepo/charmstore.go]: s/os.FileMode//

File: charmrepo/charmstore.go [hunk 3]
```diff
@@ -63,11 +69,65 @@ func NewCharmStore(p NewCharmStoreParams) (Interface, error) {
...
+	}
+	r, id, expectHash, _, err := s.client.GetArchive(curl.Reference())
+	if err != nil {
+		return nil, errgo.Notef(err, "cannot retrieve charm")
+	}
+	defer r.Close()
+
+	// Check if the archive already exists in the cache.
+	path := filepath.Join(s.cacheDir, charm.Quote(id.String())+".charm")
+	if verifyHash384(path, expectHash) == nil {
```
Comment (rogpeppe) [inline charmrepo/charmstore.go]: I wonder if we should be using sha256 for this so that we don't blow everyone's cache.

**Comment by mhilton** [comment]
:+1: LGTM thanks

File: charmrepo/charmstore.go [hunk 4]
```diff
@@ -63,11 +69,65 @@ func NewCharmStore(p NewCharmStoreParams) (Interface, error) {
...
+	// Disable bundle deployment.
+	if curl.Series == "bundle" {
+		return nil, errgo.Newf("expected a charm URL, got bundle URL %q", curl)
+	}
+
+	// Prepare the cache directory and retrieve the charm.
+	if err := os.MkdirAll(s.cacheDir, os.FileMode(0755)); err != nil {
+		return nil, errgo.Notef(err, "cannot create the cache directory")
+	}
+	r, id, expectHash, _, err := s.client.GetArchive(curl.Reference())
```
Comment (rogpeppe) [inline charmrepo/charmstore.go]: I think you should use the size here and check that the number of bytes returned by io.Copy is as expected. An error message that says there's a size mismatch, giving expected and actual sizes, is going to be much more accessible than a hash mismatch message (which of course needs to be done too).

File: charmrepo/charmstore.go (hunk 3 above)
Comment (rogpeppe) [inline charmrepo/charmstore.go]: perhaps log the error message on failure?

**Comment by rogpeppe** [comment]
LGTM with a couple of trivial suggestions, thanks!

File: charmrepo/charmstore.go (hunk 1 above)
Comment (frankban) [inline charmrepo/charmstore.go]: Removed.

File: charmrepo/charmstore.go (hunk 2 above)
Comment (frankban) [inline charmrepo/charmstore.go]: Done.

File: charmrepo/charmstore.go (hunk 4 above)
Comment (frankban) [inline charmrepo/charmstore.go]: Done.

File: charmrepo/charmstore.go (hunk 3 above)
Comment (frankban) [inline charmrepo/charmstore.go]: Done.

**Pull request merged**